src/setup/node.rs

```diff
@@ -0,0 +1,167 @@
+use fs_extra::dir;
+
+use super::constants::{ALGORAND_SETUP_DIR, NODE_DIR, PRIVATE_NETWORK_DIR};
+use crate::setup::{
+    config::{NodeConfig, NodeMetaData},
+    get_algorand_work_path,
+};
+
+pub enum ChildExitCode {
+    Success,
+    ErrorCode(Option<i32>),
+}
+
+pub struct NodeBuilder {
+    /// Node's startup configuration.
+    conf: NodeConfig,
+    /// Node's process metadata read from Ziggurat configuration files.
+    meta: NodeMetaData,
+}
+
+impl NodeBuilder {
+    /// Creates a new [NodeBuilder].
+    pub fn new() -> anyhow::Result<Self> {
+        let setup_path = get_algorand_work_path()?.join(ALGORAND_SETUP_DIR);
+
+        let conf = NodeConfig::default();
+        let meta = NodeMetaData::new(setup_path)?;
+
+        Ok(Self { conf, meta })
+    }
+
+    /// Creates a [Node] according to configuration and starts its process.
+    pub async fn start(&mut self, target: &Path) -> Result<Node> {
+        if !target.exists() {
+            fs::create_dir_all(&target)?;
+        }
+        let setup_path = get_algorand_work_path()?.join(ALGORAND_SETUP_DIR);
+
+        // Currently we can start only the first node
+        let source = get_node_path(0)?;
+
+        let mut copy_options = dir::CopyOptions::new();
+        copy_options.content_only = true;
+        copy_options.overwrite = true;
+        dir::copy(&source, &target, &copy_options)?;
+
+        // TODO(Rqnsom) configure the node
+
+        self.meta.start_args.push("-d".into());
+        self.meta.start_args.push(target.into());
+
+        if self.conf.log_to_stdout {
+            self.meta.start_args.push("-o".into());
+        }
+
+        let node = self.start_node();
+
+        // TODO(Rqnsom) wait for the connection to confirm
+
+        self.meta = NodeMetaData::new(setup_path)?; // Reset args
+        Ok(node)
+    }
+
+    /// Sets whether to log the node's output to Ziggurat's output stream.
+    pub fn log_to_stdout(mut self, log_to_stdout: bool) -> Self {
+        self.conf.log_to_stdout = log_to_stdout;
+        self
+    }
+
+    fn start_node(&self) -> Node {
+        let (stdout, stderr) = match self.conf.log_to_stdout {
+            true => (Stdio::inherit(), Stdio::inherit()),
+            false => (Stdio::null(), Stdio::null()),
+        };
+
+        let child = Command::new(&self.meta.start_command)
+            .current_dir(&self.meta.path)
+            .args(&self.meta.start_args)
+            .stdin(Stdio::null())
+            .stdout(stdout)
+            .stderr(stderr)
+            .spawn()
+            .expect("Node failed to start");
+
+        Node {
+            child,
+            config: self.conf.clone(),
+        }
+    }
+}
+
+pub struct Node {
+    child: Child,
+    config: NodeConfig,
+}
+
+impl Node {
+    pub fn builder() -> NodeBuilder {
+        NodeBuilder::new()
+            .map_err(|e| format!("Unable to create a builder: {:?}", e))
+            .unwrap()
+    }
+
+    pub fn stop(&mut self) -> io::Result<ChildExitCode> {
```

Comment:
we should comment here that `&mut self` is due to the `Drop` impl, otherwise `self` could be a more obvious pick

Reply:
Done